### PR TITLE
Resmgr list available nonexisting

### PR DESCRIPTION
**Pull request by joschrew**
Fix for https://github.com/OCR-D/core/issues/854

### CODE REVIEW

File: ocrd/ocrd/cli/resmgr.py [hunk 1]
```diff
@@ -45,7 +45,11 @@ def list_available(executable=None):
     List available resources
     """
     resmgr = OcrdResourceManager()
-    for executable, reslist in resmgr.list_available(executable):
+    resources = resmgr.list_available(executable)
+    if executable and not resources:
+        print("Error: No resources available for '%s'" % executable)
+        sys.exit(1)
+    for executable, reslist in resources:
```
Comment (bertsky) [inline ocrd/ocrd/cli/resmgr.py:52]: That hunk should be reverted for several reasons:
- we cannot just `print`, that would mix expected output (usually to be parsed by another tool) with messages (which should use logging and thus end up in a file or on stderr)
- we don't want a failure, just an empty result in that case
- it's not necessary, your fix in `OcrdResourceManager.list_available` is already sufficient

```suggestion
    for executable, reslist in resmgr.list_available(executable):
```

**Review by bertsky — CHANGES_REQUESTED** [review]
Otherwise fine – thanks!

File: ocrd/ocrd/cli/resmgr.py (hunk 1 above)
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:52]: Yes, I understand the @joschrew's intention to give a better error messages. But I agree with @bertsky this is not the right place, or rather it is not necessary.

If we want to make it explicit that there are no resources for an executable, the right place to adapt ist the `print_resources` function.

File: ocrd/ocrd/resource_manager.py [hunk 2]
```diff
@@ -84,6 +84,8 @@ def list_available(self, executable=None):
         List models available for download by processor
         """
         if executable:
+            if executable not in self.database:
+                return []
             return [(executable, self.database[executable])]
```
Comment (kba) [inline ocrd/ocrd/resource_manager.py:89]:
```suggestion
            return [(executable, self.database.get(executable, []])]
```

We need to return the same format, a list of tuples `[(executable, list_of_resources)]`, returning just an empty list is not enough.

File: tests/test_resource_manager.py [hunk 3]
```diff
@@ -104,5 +104,16 @@ def test_default_resource_dir(tmp_path):
...
 
+def test_list_available(tmp_path):
+    mgr = OcrdResourceManager(xdg_data_home=tmp_path)
+    res = mgr.list_available()
+    assert len(res) > 0
+
+def test_list_available_with_unknown_executable(tmp_path):
+    mgr = OcrdResourceManager(xdg_data_home=tmp_path)
+    res = mgr.list_available(executable="ocrd-non-existing-processor")
+    assert len(res) == 0
```
Comment (kba) [inline tests/test_resource_manager.py:115]:
```suggestion
    assert len(res[0][1])  == 0
```

**Review by kba — CHANGES_REQUESTED** [review]
Thanks for the PR, it solves the issue, just a few adaptions and then we can merge!

File: ocrd/ocrd/resource_manager.py (hunk 2 above)
Comment (bertsky) [reply ocrd/ocrd/resource_manager.py]: @kba I don't think that's true. If the returned list is empty, then nothing will get printed (which is probably what you want when using `ocrd resmgr list-available -e ...` for other tools). In your variant `ocrd.cli.print_resources` will print the name of the processor, followed by an empty line.

Comment (kba) [reply ocrd/ocrd/resource_manager.py]: Yes, that's why I meant that we should adapt `print_resources` to print something like

```
ocrd-foo-bar: NONE
```

or similar.

**Comment by joschrew** [comment]
ok, thanks for all your support. I think (hope) I changed it to that what kba initially showed me and it is working now.

**Review by kba — APPROVED** [review]

**Pull request merged**